src/mlpack/methods/kde/kde_model.hpp

```diff
@@ -0,0 +1,369 @@
+/**
+ * @file kde_model.hpp
+ * @author Roberto Hueso
+ *
+ * Model for KDE. It abstracts different types of tree, kernels, etc.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_KDE_MODEL_HPP
+#define MLPACK_METHODS_KDE_MODEL_HPP
+
+// Include trees
+#include <mlpack/core/tree/binary_space_tree.hpp>
+#include <mlpack/core/tree/octree.hpp>
+#include <mlpack/core/tree/cover_tree.hpp>
+#include <mlpack/core/tree/rectangle_tree.hpp>
+
+// Include kernels
+#include <mlpack/core/kernels/kernel_traits.hpp>
+#include <mlpack/core/kernels/gaussian_kernel.hpp>
+#include <mlpack/core/kernels/epanechnikov_kernel.hpp>
+
+// Remaining includes
+#include <boost/variant.hpp>
+#include "kde.hpp"
+
+namespace mlpack {
+namespace kde {
+
+//! Alias template.
+template<typename KernelType,
+         template<typename TreeMetricType,
+                  typename TreeStatType,
+                  typename TreeMatType> class TreeType>
+using KDEType = KDE<metric::EuclideanDistance,
+                    arma::mat,
+                    KernelType,
+                    TreeType,
+                    TreeType<metric::EuclideanDistance,
+                             kde::KDEStat,
+                             arma::mat>::template DualTreeTraverser>;
```

Comment:
What happens if you modify `KDEType` to require the `DualTreeTraverser` type as a third template parameter?  I wonder if this might help fix the Windows build.

Reply:
Doesn't work either, plus there're other places in mlpack where it's done in the same way and it works (e.g ```neighbor_search/ns_model.hpp```)

Reply:
Another idea might be to remove lots of types from the `boost::variant` type just to see if that helps.  Often with MSVC finding the actual problem involves a lot of commenting stuff out to pinpoint the problem...

Reply:
I think I just found the error. Apparently this kind of partial specialization is not allowed by MSVC.
```
  //! DualBiKDE specialized on Gaussian Kernel KDEType.
  template<template<typename TreeMetricType,
                    typename TreeStatType,
                    typename TreeMatType> class TreeType>
  void operator()(KDETypeT<kernel::GaussianKernel, TreeType>* kde) const;
```
I can either fully specialize KDETypeT or not specialize at all.

Reply:
Hmm, you might be able to sidestep the specialization a bit by not specializing on the kernel type and instead, inside the implementation of `operator()`, do something like...

```
if (std::is_same<KernelType, kernel::GaussianKernel>::value)
{
  // Gaussian kernel implementation...
}
else if (std::is_same<KernelType, ...>::value)
{
  ...
}
```

That might be kind of ugly but it would probably at least work on Windows.  What do you think?

Reply:
The problem with that implementation is that for some kernels it calls ```kernel.Normalizer()``` and since not every kernel has that method, it won't compile either :(

Reply:
Maybe you could do something like this?

```
template<typename KernelType>
void DoSomethingWithKernel()
{
  // default implementation
}

template<>
void DoSomethingWithKernel<GaussianKernel>()
{
  // for a kernel that has Normalizer()...
}

template<typename KernelType,
          template<typename TreeMetricType, typename TreeStatType, typename TreeMatType> class TreeType>
void operator()(KDETypeT<KernelType, TreeType>* kde) const
{
  // We don't specialize KernelType here, but instead call out to the correctly specialized version of
  // DoSomethingWithKernel<>()...
  ... DoSomethingWithKernel<GaussianKernel> ...
}
```

That's not tailored to the code you're writing but hopefully you get the idea---we can use a generic `operator()` which in turn calls out to some function that is specialized for certain kernel types.  I think this strategy could work as a workaround... let me know if I overlooked something.

Reply:
Thanks for the idea! :) Solved in 170039423f97a2b065998ac330422ffa1fd11962